Add support for templating node labels and annotations

solidDoWant opened the pull request:
This adds support for templating node labels and annotations. This is useful for passing Talos information that is node-specific to Kubernetes workloads. Each field will be provided the built, but not yet validated or templated machine config struct as the root context (`.`).

Here's a simple demo (included in the example file):
```yaml
# Source
nodeLabels:
  rack: rack1a
  zone: us-east-1a
  secureBootEnabled: >-
    {{
      .MachineConfig.MachineInstall.InstallImage |
      contains "installer-secureboot"
    }}
nodeAnnotations:
  rack: rack1a
  installerUrlCode: '{{ .MachineConfig.MachineInstall.InstallImage }}'
```

```yaml
# Rendered result
  nodeLabels:
    rack: rack1a
    secureBootEnabled: "true"
    zone: us-east-1a
  nodeAnnotations:
    installerUrlCode: factory.talos.dev/installer-secureboot/67a999813ff49ba00dc632aa48170dc866bda646b7ee4f944e2fb7dc50b522bd:v1.7.1
    rack: rack1a
```

I've configured the template engine to use [Sprig functions](https://masterminds.github.io/sprig/), which is what other popular Golang templating tools use (like [Helm](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/), [gotemplate](https://coveooss.github.io/gotemplate/), [gomplate](https://docs.gomplate.ca/)). I think this will cover the majority of user's needs.

If desired, it should be pretty trivial to add templating support to other fields as well. Personally I'd find a lot of value in having support for templating in `extraKernelArgs`, but I thought it'd be best to start small with labels and annotations, and get some feedback on the concept first.

Prior to merge, I need to add:
* Tests
* Docs
* ~Example~

solidDoWant commented on pkg/config/validator.go:
```diff
@@ -280,8 +280,17 @@ func checkNodeHostname(node Node, idx int, result *Errors) *Errors {
 
 func checkNodeLabels(node Node, idx int, result *Errors) *Errors {
 	if node.NodeLabels != nil {
+		// Skip labels that include templates as these will not be valid
```
Similar logic may be needed for annotations... not sure. The example works just fine, but there may be some case where templates can fail, but the rendered value would not.

budimanjojo replied:
You can do this for annotations too now?

solidDoWant replied:
Yep I added support for both. I can pull this part out if you'd like. Annotations can be a useful alternative because there is a short limit (63 characters) for label values.

budimanjojo commented:
Thanks! The one thing that I don't like about this is the skipped validation for the templated value and the only way to satisfy my dislike is to call the `labels.Validate` function again on those templated value, which looks ugly because the validation step is passed a long time ago. And there's no neat way to do it because the template data is from the already generated `talos.Provider` which is already near the end of the program.

solidDoWant commented:
> the skipped validation

While this is skipped, the tool will fail with a validation error after the config has been generated. For example, given the following config:
```yaml
nodeLabels:
  installerUrl: '{{ .MachineConfig.MachineInstall.InstallImage }}'
```

The tool will fail, prior to generating configs, with:
```console
2024/11/12 22:05:13 failed to generate talos config: 1 error occurred:
	* invalid machine node labels: 1 error occurred:
	* label value length exceeds limit of 63: "factory.talos.dev/installer-secureboot/67a999813ff49ba00dc632aa48170dc866bda646b7ee4f944e2fb7dc50b522bd:v1.7.1"
```

This is less descriptive than the pre-generation validation, but it is still _some_ form of validation.

budimanjojo commented on pkg/talos/nodeconfig.go:
```diff
@@ -53,6 +56,13 @@ func GenerateNodeConfig(node *config.Node, input *generate.Input, iFactory *conf
 
+	// Templating should be done as late as possible to maximize the amount of infomation
+	// available for templating
+	cfg, err = templateConfig(cfg)
```
I like number 2 approach. Thanks!

budimanjojo commented:
> While this is skipped, the tool will fail with a validation error after the config has been generated.

Ah yeah that's validation from the Talos side of `Validate()` function at the end of the program.

solidDoWant commented on pkg/talos/nodeconfig.go:
```diff
@@ -126,6 +138,106 @@ func applyNodeOverride(node *config.Node, cfg taloscfg.Provider) taloscfg.Provid
+}
+
+// Returns true if the provided string contains a template, false otherwise
+func isTemplateString(str string) bool {
+	return strings.Contains(str, "{{")
+}
+
+func templateMap(templateText map[string]string, data any) (map[string]string, error) {
+	// Track the new values separately so that the output is always deterministic.
+	// Otherwise, because map iteration order is random, multiple runs could
+	// produce different results in the case of referential values.
+	templatedText := make(map[string]string, len(templateText))
+	for key, templateValue := range templateText {
+		renderedValue, err := renderTemplate(templateValue, data)
+		if err != nil {
+			return nil, err
+		}
+
+		templatedText[key] = renderedValue
+	}
+
+	return templatedText, nil
+}
+
+// Takes in a Go template string and renders it.
+func renderTemplate(templateText string, data any) (string, error) {
+	builtTemplate, err := template.New("template").Funcs(sprig.FuncMap()).Parse(templateText)
+	if err != nil {
+		return "", err
+	}
+
+	buf := new(bytes.Buffer)
+	err = builtTemplate.Execute(buf, data)
+	if err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
```
It might be beneficial to move this logic (or most of it) to another package. What do you think?

budimanjojo replied:
Yeah you're right.

solidDoWant commented on pkg/talos/nodeconfig.go:
```diff
@@ -126,6 +138,106 @@ func applyNodeOverride(node *config.Node, cfg taloscfg.Provider) taloscfg.Provid
 }
 
+// Template supported fields within the config
+func templateConfig(node *config.Node, cfg taloscfg.Provider) (taloscfg.Provider, error) {
+	// Two separate copies of the configuration are required: one that is
+	// continuously updated with rendered values, and one that that contains
+	// the constant, raw template values. This ensures consistent results in
+	// the event that support for new fields are added, or that they are
+	// templated in a different order, which is important to prevent breaking
+	// changes.
+	renderedConfig := cfg.RawV1Alpha1()
+	unchangedConfig := renderedConfig.DeepCopy()
+
+	err := templateConfigField(node.NodeLabels, &renderedConfig.MachineConfig.MachineNodeLabels, unchangedConfig, "node labels")
+	if err != nil {
+		return nil, err
+	}
+
+	err = templateConfigField(node.NodeAnnotations, &renderedConfig.MachineConfig.MachineNodeAnnotations, unchangedConfig, "node annotations")
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
```
This could alternatively be written as:

```go
func templateConfig(node *config.Node, cfg taloscfg.Provider) (taloscfg.Provider, error) {
	// Two separate copies of the configuration are required: one that is
	// continuously updated with rendered values, and one that that contains
	// the constant, raw template values. This ensures consistent results in
	// the event that support for new fields are added, or that they are
	// templated in a different order, which is important to prevent breaking
	// changes.
	renderedConfig := cfg.RawV1Alpha1()
	unchangedConfig := renderedConfig.DeepCopy()

	fields := []struct{
		name string
		srcKeyPairs map[string]string
		dstKeyPairs *map[string]string
	}{
		{
			name: "node labels",
			srcKeyPairs: node.NodeLabels,
			dstKeyPairs: &renderedConfig.MachineConfig.MachineNodeLabels,
		},
		{
			name: "node annotations",
			srcKeyPairs: node.NodeAnnotations,
			dstKeyPairs: &renderedConfig.MachineConfig.MachineEnv,
		},
	}

	for _, field := range fields {
		err := templateConfigField(field.srcKeyPairs, field.dstKeyPairs, unchangedConfig, field.name)
		if err != nil {
			return nil, err
		}
	}

	return cfg, nil
}
```

The benefit here is that it's _extremely_ simple to add support for more fields - just add a new struct instance and you're done. The downside is some more complexity and abstraction. Do you have a preference?

budimanjojo replied:
I'm fine with both actually. There might be other types outside of `map[string]string` that need to be templated right? The alternative seems to be easier to work with converting type to `map[string]string`?

solidDoWant replied:
> There might be other types outside of map[string]string that need to be templated right?

I think this can be broken down into two parts: the value to be templated, and it's structural relationship with other data.

Templates must always be strings, but they could be converted as appropriate to other destination types. For example a template to calculate `bool` would always be a `string`, but could produce a truthy string (`true`/`false`) that should be converted to a `bool`.

Due to the nature of the config, each value (templated or not) must always be associated with a string-type key. In some cases (arrays), multiple values will be associated with a single key. 

With this in mind, the source type for a block of data will always be (after flattening/moving a "bottom" node in the config tree) either `map[string]string`, or `map[string][]string`. However, the destination can only be as specific as `map[string]any` and `map[string][]any`, as the resultant type may need to be something other than a string.

I think all cases could be covered by:
* Changing the destination type from `map[string]string` to map[string]any`, then adding some logic to convert the rendered value to each resultant literal type (rejecting `interface{}` types with an unsupported error)
* Adding similar logic for `map[string][]string`/`map[string][]any`

Does this approach work? If so, I'll pull all (or most) of the templating code into a separate package and then start working on this.

budimanjojo replied:
Thanks! Just do what you seem fit. I need to see the code to know how that works because I'm bad at understanding words :D

solidDoWant replied:
I worked on this quite a bit and was unable to find a way to generally describe the destination key pair types due to language constraints. I was able to make a couple of functions more generic (so that rendered types other than `string` are supported), and also added functions for templating slices.

budimanjojo commented on pkg/templating/template.go:
```diff
@@ -0,0 +1,114 @@
+// Given a map `src`, return two maps where the first contains key-value pairs that are not
+// templates, and the latter containers only key-value pairs that are templates.
+func SplitTemplatedMapItems(src map[string]string) (map[string]string, map[string]string) {
+	nonTemplateKeyPairs := make(map[string]string, len(src))
```
I think the size of the map can be omitted here?

solidDoWant replied:
It can be, but it's generally faster to specify the size if it's known ahead of time. This avoids overallocation (often by a significant amount) later, or resizing.

budimanjojo commented on pkg/templating/template.go:
```diff
@@ -0,0 +1,114 @@
+// templates, and the latter containers only key-value pairs that are templates. Values of the
+// opposing types will be set to an empty string, to preserve the position of all items.
+func SplitTemplatedListItems(src []string) ([]string, []string) {
+	nonTemplateKeyPairs := make([]string, len(src))
```
The name of the variable should be `nonTemplateList` or something else?
And the size of the slice can be `0, len(src)`? This might be more efficient? I don't know for sure though.

solidDoWant replied:
> The name of the variable should be `nonTemplateList` or something else?

Yep, you're right. Copy/paste error. Made a similar mistake below as well.

> And the size of the slice can be `0, len(src)`? This might be more efficient? I don't know for sure though.

Normally yes but in this specific case no. The second argument is the length, and the last is the capacity. The goal here is to produce two arrays of the same _length_, with nil/default/zero values for unwanted items. This preserves the order of items in the list so that they can later be combined together again, after templating.

budimanjojo commented on pkg/talos/nodeconfig.go:
```diff
@@ -99,13 +107,15 @@ func applyNodeOverride(node *config.Node, cfg taloscfg.Provider) taloscfg.Provid
 	if node.NodeAnnotations != nil {
-		slog.Debug(fmt.Sprintf("setting node annotations for %s", node.NodeAnnotations))
-		cfg.RawV1Alpha1().MachineConfig.MachineNodeAnnotations = node.NodeAnnotations
+		nonTempalteAnnotations, _ := templating.SplitTemplatedMapItems(node.NodeAnnotations)
```
typo in variable name?

solidDoWant commented on pkg/templating/template.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// Returns true if the provided string contains a template, false otherwise
+func IsStringATemplate(str string) bool {
```
This is more philosophical than technical, but should the following strings be considered malformed templates, or non-templates?
* `" .Node.Name }}"`
* `" {{ .Some.Value "`
* `"}} {{ .AnotherValue }}"`

I am of the opinion that if a string contains `{{` or `}}` (I missed this one), then it's probably better to assume that the user meant to use a template. There aren't a lot of cases where these values would be desired/used outside of templating.

What do you think? I'll defer to your opinion.

solidDoWant replied:
See [here](https://github.com/budimanjojo/talhelper/pull/701/files#diff-9a39b1c30a3713cf232bac75ceb20106f7a47873969e9c2440cc246393045343R113-R168) for some test cases I wrote covering this

budimanjojo replied:
I think using contains `{{` or `}}` will returns true for `only-opening`, `only-closing`, and `out-of-order` in your test? Because we should check if the index of first `{{` is there before the index of first `}}` and a string that only contains opening or closing is not a template?

budimanjojo replied:
But you're right in a sense where it should error out in the templating function instead of being processed as a non-template and pass the labels/annotations validation.

solidDoWant replied:
So just to make sure we're on the same page, any string containing one or more `{{` or one or more `}}` should be considered a template string, right? And if they `{{`/`}}` are in the wrong place, aren't paired, or are otherwise invalid, then they should still be considered a template string, which the templating library will reject, right?

budimanjojo replied:
Yeah it makes sense. The downside is if the user really want to put something like only-opening, only-closing, or out-of-order scenario as value, I don't know if there's any use case that requires that but we can improve that in the future when that scenario happens IRL. Maybe you should also add a code comment about this quirk in case the future people looking at this code got confused.

solidDoWant commented:
Added tests in https://github.com/budimanjojo/talhelper/pull/701/commits/fc1a8b80b375cd1c1743af1d53aabad9eb97419e. How/where would you like documentation added? I'm not quite sure which pages should have this information added, how it should be formatted, etc.

> Also can you squash the commit as `feat(genconfig): support templating node labels and annotations`?

I'll be sure to squash and rename after final review.

solidDoWant commented:
Added docs. Let me know what you think, and if everything is good to go, then I'll squash and fixup the commit message.

I added a link to my system-upgrade-controller deployment in the docs, as a full example of how to handle automated Talos upgrades. A full example (including scripts, health checks, scheduling, etc.) would require a bunch of page space so I thought it might be better to just show a small portion of the concept, with a link to the full thing. Not sure what your policy on this is - let me know if you want me to remove it.

budimanjojo commented:
Looks good! I'm running the workflows now, let me know if you're done with the commit squashing. Thank you!

budimanjojo commented:
v3.0.10 released with this feature!
